### ISSUE TITLE
traverse/transform: path.remove() or any other way to change the collection does not work

### ISSUE THREAD

**Issue by hg42** [DmitrySoshnikov/regexp-tree#37]
Hi,

Changing the parent collection (e.g. expressions) in a traverse() or transform() has an effect on the traversion.
E.g. the next node will be skipped if I use this (in coffeescript):
```coffeescript
    RegExpTree.traverse(ast, {

      Char: (path) ->
        {node, parent, property, index} = path
        if parent.type == "Alternative"
          if node.value == " " or node.value == "\n"
            path.remove()
      }
    )
```
I assume, the index is used in the iteration and that changes when removing the item from the collection.

Perhaps I am doing it wrong? but I assume the remove() of NodePath is for exactly this purpose.
What is the preferred way to delete a node inside traverse()?

As a workaround I tried to set `node.type = null` or to an undefined type name, but this gives me an error.
Would be nice if traverse() and generate() could skip such a node.
Also optimize could remove such nodes.

**Comment by hg42** [comment]
I can also look at it myself and create a PR?

**Comment by DmitrySoshnikov** [comment]
@hg42, good catch, thanks! Yeah, the traversal tracks, and [passes](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/traverse/index.js#L89) the index in the original AST. Once we removed a node from a handler (when a handler is [executed](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/traverse/index.js#L108)), it seems the original, and actual indexes are getting out of sync.

I can take a look later into this, but also will definitely appreciate a PR for this in case you'd like to tackle it. The traversal/transform APIs are in development at the moment, so issues like this one might still arise, and great that they are found.

In some edges cases one can use [direct nodes manipulation](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/optimizer/transforms/char-star-to-plus-transform.js#L40-L43), however, having it fully covered by the transform API methods is much better for sure.

**Comment by hg42** [comment]
I believe `parent[path.property].splice(path.index - 1, 1);` will also trigger this issue, but the test probably doesn't catch that, because there is nothing following.

I first used `parent[path.property].splice(path.index, 1)` to remove the node, but this didn't work and that's why I looked for other ways and found one in NodePath, which btw. uses exactly the same method
`this.parent[this.property].splice(this.index, 1);`

**Comment by hg42** [comment]
btw. I forgot to say your package is really awesome :-)
It solved several problems I had with javascript RegExp in a single strike by transforming it.

**Comment by DmitrySoshnikov** [comment]
> I believe parent[path.property].splice(path.index - 1, 1); will also trigger this issue

Yeah, absolutely correct :( we'll need to fix this.

And thanks, glad the tool is useful to you!

**Comment by hg42** [comment]
I looked into the code and found that it's not possible to fix this without changing **ast-traverse**.

It's generally risky to edit a collection in place while iterating over it.

Probably, many operations of an optimizer will not work this way, especially when iterating using an index (like ast-traverse does).

Usually, I would use linked lists for that (coming from C-like languages or may be LISP), then you can insert and delete by relinking the list (or the tree!). Often you get the next element before changing the tree. And because it's a single object, the reference to it is still valid after reordering the collection it belongs to.

But I'm not up-to-date on that topic, there might be other useful design patterns working on javascript like structures.

May be it's better to implement transform as regenerating the AST. The result of `visit()` would be added to the current collection.

I would also prefer to always have the same `node.expressions` array instead of different variants (`node.expression` vs. `node.expressions`). This would allow to transparently traverse the (abstract) tree without knowing anything about it's nodes. `prop` and `index` would not be necessary then.

**Comment by hg42** [comment]
For now I have set `node.type = "None"` and run a cleanup afterwards.

One "solution":
```coffeescript
      any = 1
      while any
        any = 0
        RegExpTree.traverse(ast, {
          None: (path) ->
            any = 1
            path.remove()
        })
```
I'm not sure if this could fail in some way because the iteration may get to a non existing array item (because it was shifted away).
Would be better to traverse all collections and cleaning up the expressions array.
However, you would have to know which classes exist and which are collections and you would have some code duplication. So it's additional work whenever a relevant node type is added.

Oh wait...I just found a **catch all handler** (= "*") which allows to traverse without such knowledge:
```coffeescript
    RegExpTree.traverse(ast, {
      "*": (path) ->
        exprs = path.node.expressions
        if exprs
          show path.node.type + ": " + exprs.length
          filtered = []
          for node in exprs
            if node.type != "None"
              filtered.push node
          path.node.expressions = filtered
    })
```
and this works.

This might be a candidate for a standard function of the package.
When traversing like this, you would not need to use a named type like "None" or may be "Delete", you could also set it to null, which is more natural.

Eventually the cleanup could be part of the visitor loop (assuming ast-traverse would be incorporated into the package). The visitor could delete the item when it's type is null after the visit and then adjust the index.
May be this would be a good solution.
But anyways, it doesn't help with bigger transformations like `aa* -> a+`.
Deleting the current node is only one of many possible restructuring operations.
E.g. flattening a (sub-)tree would dramatically shift the array items of the base level.
I think, a simple iteration like in ast-traverse cannot handle this.

**Comment by DmitrySoshnikov** [comment]
Thanks for the analysis, yeah, all points are legit. I'll have to think about the modifications, and as mentioned previously, probably we'll need to remove dependency on `ast-traverse`, and build a custom version.

Re: `node.expression` vs. `node.expressions`. This is mostly a semantic difference. E.g. a group has an `expression`, since it only can have 1 expression, hence "s" suffix semantically doesn't fit. I was thinking to have generic `body` instead of `expression`/`expressions`, this might solve the issue. The `property` will still be needed perhaps (in order to delete a node from a parent one).

**Comment by DmitrySoshnikov** [comment]
Oh, nice trick with "catch-all" `*`! It still has two-pass of the AST, but can solve the issue. Anyways, I'll think more about these edge-cases, and you are right, more complex operations will be harder with the current issue of unsync "index".

**Comment by hg42** [comment]
it may be even worse because the index runs up to the length which is probably cached by the for loop (not sure).
So if the array would become longer the loop would stop before end.

You could perhaps look how big engines like LLVM do such transformations. I guess they try to be very efficient, though javascript is not C.
There might be other projects like coffescript or other precompilers that have good solutions to copy from.

**Comment by DmitrySoshnikov** [comment]
Yeah, at least the length [isn't cached](https://github.com/olov/ast-traverse/blob/master/ast-traverse.js#L28), so this should work in general, but I'll think about restructuring this anyways.

**Comment by DmitrySoshnikov** [comment]
OK, this should be fixed now in https://github.com/DmitrySoshnikov/regexp-tree/commit/9859d991bd20f98f863ca6559dbd932a152a1d69 (and in `0.0.28`). Please let me know if you see any issues. I also exposed `getNextSibling`, and `getPreviousSibling`, which return paths now.

**Comment by hg42** [comment]
ok, first I thought it looks good to me. But in my simple usage (deleting a space char) this doesn't work.
Seems the tests are not covering this (though I first thought they would).
That's because you only delete one node and test only on the NodePath level.
You can detect this if you create a regexp `/abcd/`, delete character `b` and `c` in `traverse()` and look if the generated regexp is `/ad/`. You should find the result is `/acd/` because `c` was skipped.

That's because, what I said before about ast-traverse is still valid:

ast-traverse runs an index. You are changing indexes at NodePaths, but you don't have access to the loop index of ast-traverse.

Say it is at index 5 when you come into the game.
You delete yourself at index 5.
Now the next index must be 5, because your next sibling is shifted to index 5.
But ats-traverse increments it's current counter (5) and continues with index 6 => wrong.

I think, you can only solve this with changes to ast-traverse (or implement it yourself).

May be the loop_index should be part of the NodePath the visitor is working on (that one where the array lives). So it could be corrected by `remove()`.

That's why I didn't try to provide a PR, there are too much architectural decisions to be made.

Note, optimizations often go over several layers, most cases will probably reorder a tree, not only a linear structure. You may change indexes of an outer visitor loop (on a lower recursion level).
So solving this in the current level would only be a first step.

However I'm not sure if regexps fall in such categories.
May be regexp optimizations never get that complicated. More simple things like removing empty parens come to my mind.
What are you aiming at with `optimize()`?

**Comment by hg42** [comment]
I am thinking of something like this in ast-traverse:
```js
    for (node.loop_index = 0; node.loop_index < child.length; node.loop_index++) {
        visit(child[node.loop_index], node, prop, node.loop_index);
    }
    # may be remove node.loop_index here which also indicates that a loop is not running
```
and
```js
    parentPath.loop_index--;
```
in node-path.js

**Comment by DmitrySoshnikov** [comment]
@hg42, oooh, great catch again, thanks!

> You can detect this if you create a regexp `/abcd/`, delete character b and c in traverse() and look if the generated regexp is `/ad/`. You should find the result is `/acd/` because c was skipped.

Yes, I'll make sure this test is passing, and yes, perhaps we'll have to remove `ast-traverse` still, will check.

Also a good point on more complex optimizations which can reshuffle several layers, though I hope we'll do gradual fixes if we find though issues.

> More simple things like removing empty parens come to my mind.
What are you aiming at with `optimize()`?

Empty parens is a good use-case BTW, we should add it. The list of transformations we have in mind yet for optimize is [this one](https://github.com/DmitrySoshnikov/regexp-tree/issues/21#issue-218097307). Not limited though, feel free to add anything there.

**Comment by DmitrySoshnikov** [comment]
@hg42, I added the test for removal of 2 characters, and seems it's passing. Please feel free to review the https://github.com/DmitrySoshnikov/regexp-tree/commit/35688985821fa90045e3a3dafc3d45e871dee64c. Does it covers the use-case?

If though I'm missing the actual use-case (might be), can you provide an actual code snippet you use for traversal, and removal?

(I should probably start doing pull-requests in case some one will be interesting to actually reviewing them)

**Comment by hg42** [comment]
nope, these lines don't trigger the problem:
```js
    // Remove 'b' itself.
    path.remove();
    // Remove 'c' as well.
    cPath.remove();
```
because after deleting both nodes from the array processing is finished, so the wrong index in ast-traverse doesn't matter any more.
The problem only appears when you delete a node and return to ast-traverse and want to process a next node.
So, you should change it to:
```js
    if(value == "b")
        path.remove();
    else if(value == "c")
        path.remove();
```
question is how to check expectations.
You already have a second traverse which checks if "b" and "c" are removed.
This may be a matter of taste, but IMHO, all this code hides the intention of the test.
Many things are already tested in another test.
The simplest check would use generate() and expect a result.
Or if you do not want to use generate() you could check values of AST nodes instead, e.g. you could look for "a" and "d" at the expected positions.

**Comment by DmitrySoshnikov** [comment]
Cool, thanks, I'll look into this example, and will provide a smaller test with "transform-generate" (we can keep this one detailed test, I wanted to explicitly track index changes).

**Comment by DmitrySoshnikov** [comment]
OK, I investigated it, and while it is possible to adjust traversing index during traversal (by checking length of the collection before, and after call, calculating delta, and fixing `i`), I think a better approach could be a "post-cleanup" similar to what you had with "None" type.

In this case removing a node will mean not fully remove it from the array, but marking as `null`. And post-traversal of the collection, filter the `null` values.

Thus the original index of the node will be preserved during transformation.

```
['a', 'b', 'c', 'd', 'e']

// Removed 'b', and 'c'
['a', null, null, 'd', 'e']
```

The index of `d` will be still 3 in this case. 

```
// Inserted 'z' at 1, it won't be automatically visited
// and if needed, a caller should explicitly call visit(...) API method
// which will expose.
['a', 'z', null, 'd', 'e']
```

If node is `null` it's not visited (e.g. you remove 2 nodes, and next traversing index will get `null` as a node, and should not call a handler).

Once a collection is finished, do a clean up with adjusting index in `NodePath` instances:

```js
['a', 'z', 'd', 'e']
```

Thoughts?

**Comment by DmitrySoshnikov** [comment]
Probably even won't remove `ast-traverse`, since it [checks](https://github.com/olov/ast-traverse/blob/master/ast-traverse.js#L10) `if (!node ...) { return; }`. So "removed", `null` nodes won't be just visited by other visitors. Still good to do a clean up though.

**Comment by DmitrySoshnikov** [comment]
There might be problems with insertions though. We don't have an API to insert nodes yet, but it's possible doing "raw manipulations", and index may change when you insert the nodes. In this case it should be fine though, and nodes won't be skipped, since `length` is not cached.

**Comment by DmitrySoshnikov** [comment]
OK, after rethinking again, I think we should stick with actual index recalculation in order to cover insertion of a node. In this case removal/insertion will be handled generically, and will track correct index.

**Comment by DmitrySoshnikov** [comment]
@hg42, OK, in https://github.com/DmitrySoshnikov/regexp-tree/commit/214ef518ed9ed49436c132cadaad757949bea55c I used a similar to yours proposed approach with `loop_index`, and adjust current `i` index based on the tracked removed indices (one handler may remove several nodes).

I also removed dependency on `ast-traverse`, so we control now the traversal, and can adjust indices as needed. This will be needed when we'll introduce `insert` API as well.

Let me know if you still see any issues!

**Comment by hg42** [comment]
wow, you are really fast :-)
I just wanted to write all those points you say above (insert etc.), and also created a test case nearly identical to yours.
So nothing to do for me :-) just agree.
I'll test it soon.

**Comment by hg42** [comment]
> OK, after rethinking again, I think we should stick with actual index recalculation in order to cover insertion of a node. In this case removal/insertion will be handled generically, and will track correct index.

ok, you collect the removedIndices to check them if they affect the current loop index.
But to my intution these lines look suspicious to me, because you change the index you compare with:
```
            if (NodePath.removedIndices.length > 0) {
              NodePath.removedIndices.forEach(index => {
                if (index <= i) {
                  i--;
                }
              });
```
at least I cannot see correctness at first sight.

But this seems to be correct on second sight, because you simply seem to delay the influences of the removes on the loop index by keeping a chronological list of actions and executing them later in the same sequence after the loop.
Did you design it this way intentionally?

When later expanding this to allow insertion, it will get even more complicated. You would have to store the actions like {remove: 1}, {insert: 5}. You cannot keep two different array for removals and insertions, because the sequence matters.

Anyways, I think `traverse()` should be a method of NodePath.
The main advantage is, that all functional dependencies would be local to NodePath, implementation detailsare hidden and it's more guarantied that changes stay in sync.
Also, you can naturally express traversal of a subtree as `path.traverse(...)`.

A global traverse can still exist but redirect to NodePath's `traverse()`.
I understand, if you want it as a module.
You can either let the module extend the NodePath API (but then you still have different modules that are highly dependent on each other) or you can simply declare traversal as a base functionality of a tree (which it is IMHO) and eliminate the traverse module.

**Comment by hg42** [comment]
it works for my use case.

Because I didn't expect this to work, I was going to suggest something about storing a path from get*Sibling() and removing it via the stored path.

But I see you already have a test exactly doing this (I started to write one:-).
So this works already.
I'm impressed, very well done!

This works because the "stored" path is only a reference into the array, right?
So, if you would make a deep copy this would get out of sync.
So, I wonder what happens, if you do
```js
path.remove()
path.remove()  # logically illegal, but should do nothing, because this reference is already removed
```
you seem to have thought about this already, setting this.index and this.property to null:
```ja
    if (this.index !== null) {
      ...
      this.index = null;
      this.property = null;
      return;
    }
    // A simple node.
    delete this.parent[this.property];
    this.property = null;
```
but I believe you still have to guard the last two lines with `if(this.property !== null)`?
Or may be set `this.parent=null`, because it's checked early?

What do you want to allow someone to do with a removed path?
I would either expect it to be totally illegal to use it after removal.
Or all operations not using the index could be still possible (even then a bad strategy I guess).
`get*Sibling()` or other operations depending on the path.index don't make sense any more.
Access to `path.parent` could be allowed.
`path.property` could also be useful and could still be left intact.
Even `path.node` could still be valid (while hanging around without connection).

I think, it would be a better strategy to make the path totally unusable, may be even through an exception.
The user can store what he needs to use after the removal.
It's easier to document/explain/notice this simple rule and you are more free to implement internals differently.

**Comment by hg42** [comment]
>> at least I cannot see correctness at first sight

perhaps a comment to state that it's a kind of journaling system could be helpful.
You could use `journal` or `actionJournal`  instead of `removedIndices` (when adding `insert()`) to make the intention more clear.

Instead of `{remove: <index>}` and `{insert: <index>}` you could use positive indices for insertion and negative for removal if you want to keep it slim and fast. But this is less explicit and not expandable to more kinds of actions. But may be there are no more kinds of actions to be added to the journal, at least I don't imagine any.

**Comment by DmitrySoshnikov** [comment]
@hg42, cool, thanks!

> these lines look suspicious to me  ... But this seems to be correct on second sight ...
Did you design it this way intentionally?

I actually thought I had an error there as well because of `i--` modification :) But when checked, it should work, even if you delete a previous node, and itself, the `if (index <= i)` should still stand. We can change it to be more clear, like:

```js
let numberToDecrease = 0;
...
if (index <= i) {
  numberToDecrease++;
}
...
// And later:
i -= numberToDecrease;
```

But seems `i--` "just works" too.

> You would have to store the actions like {remove: 1}, {insert: 5}. You cannot keep two different array for removals and insertions, because the sequence matters.

Actually, you are right, I'm gonna change it then to this approach (using `actionJournal`) when we'll be adding `insert` operation.

> Anyways, I think traverse() should be a method of NodePath.

Yes, I thought about it as well. The `traverse` (or `visit`) method will be needed on a `NodePath` anyways when we'll have to handle some edge case, like nodes insertion in some deeply previous positions in a tree, and those node won't be visited (so callers will have to explicitly visit them).

> A global traverse can still exist but redirect to NodePath's traverse()

Yeah, it might be still good to have the `traverse` module by itself. Currently it also supports "raw nodes" traversing mode (when no `NodePath` wrapper is created). Not sure if we still need to keep this, or should deprecated now -- since not that much you can do with such a traversal -- node removals, insertion won't work properly if you do manual modifications. Though it still might be useful for simple traversals of a tree for some purposes, when no modifications are needed.

> This works because the "stored" path is only a reference into the array, right?
So, if you would make a deep copy this would get out of sync.

Correct, a `NodePath` for a node exists in one exemplar, and is taken from the [registry](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/traverse/node-path.js#L151). Cloning would cause the issues I guess, but I'm not sure when we'll need to clone one. Inserting a new node will just create a new `NodePath`, and store it in the registry.

> So, I wonder what happens, if you do

```js
path.remove()
path.remove()  # <- logically illegal, but should do nothing,
```

Ah, should probably be a "no-op". A `NodePath`, has [isRemoved](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/traverse/node-path.js#L84) predicate, which is also [used at traversal](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/traverse/index.js#L183-L185) to determine whether a next handler should be called for this path. So I think we should check, and just return from `remove` in this case.

> but I believe you still have to guard the last two lines with `if(this.property !== null)`

Yeah, it might be good to have for a double-protection, though if we early return in case of `isRemoved`, this should work as is.

> What do you want to allow someone to do with a removed path?
I would either expect it to be totally illegal to use it after removal.

That's said, the only operation on it is checking whether it's removed, via `isRemoved`, and which  traversal uses. There should be no any other operations allowed, that's true (so we should do `isRemoved` check in all modifying methods).

> You could use `journal` or `actionJournal` instead of `removedIndices`

Yep, we should switch to it once `insert` is introduced.

> Instead of {remove: <index>} and {insert: <index>} you could use positive indices for insertion and negative for removal if you want to keep it slim and fast. But this is less explicit and not expandable to more kinds of actions. But may be there are no more kinds of actions to be added to the journal, at least I don't imagine any.

Yeah, I like explicit `{remove: <index>}`, and `{insert: <index>}` approach. It allows adding more actions in the future if needed. Lower-level encoding is faster for sure, but it's fine in this case I think to have full `{remove: <index>}`, etc.

P.S.: (BTW, I assume you're in Germany, so there might be some 12 hours delay in my responses/commits :) und die deutsche Sprache gefällts mir)

**Comment by DmitrySoshnikov** [comment]
OK, closing this issue, feel free to add any comments or reopen in case if issue with `remove` will be found. I open a separate issue #45 to implement `insert` operation.

**Comment by hg42** [comment]
I thought of this before, but now I think, *this* would *NOT* work:
```js
let numberToDecrease = 0;
...
if (index <= i) {
  numberToDecrease++;
}
...
// And later:
i -= numberToDecrease;
``
Correctly, you would decrement the loop index immediately when removing the item.
For the next remove (or insert) operation the index would be lowered by one (hope my English works :-).
I think it's correct to check if the index is below the current index, because otherwise the current index would not be affected (the removal comes after the current loop index).

**Comment by hg42** [comment]
simple example:

you have `/0123456789/`, i is 5, so it points to "5".
you remove  3 times at index 0 ("0", "1", "2"), so you have `/3456789/`, numberToDecrease would be 3.
now you remove three times at index 4 ("7", "8", "9") giving `/3456/`,
4 <= 5 for all three, so numberToDecrease is now 6.
after all i is -1 (=5-6) (negative!).

If you compare with the actual loop index you have i at 5-3 = 2 after the first three removals.
For removals at 4 you have 4 not <= 2, so i stays at 2 for all three.
So it points to "5", which is correct.

**Comment by DmitrySoshnikov** [comment]
doh! yeah, so it's a tricky problem actually, when a solution is not seen right from the beginning. So the simple `i--` then works, even if it seemed it won't work.

**Comment by DmitrySoshnikov** [comment]
FYI: in https://github.com/DmitrySoshnikov/regexp-tree/commit/3e537a42923407b0c99144bcf61817a35755b3ac I added `insertChildAt` operation, and found that tracking insert/remove with `i--` became really hard. So I rewrote it with a current traversing index, which is used by `traverse`, and can be update in remove/insert methods.

**Comment by hg42** [comment]
Nice, will have a look...

Btw. I had the desire to revisit the newly created items. That's why I made a second pass.

For general algorithms it may be necessary to decide if I want to have my handler executed before or after the children (or even both). I think astTraverse has pre and post for the same reason.

Changing the node type may skip another handler which you thought to handle that.
Generally changing the tree while traversing it increases complexity.
You have many things to think of. Some of which are not clearly visible or implementation details.
You can reduce that complexity with more passes.

This is also very much related to the need for NodePaths.
To ensure being able  to do all necessary operations (navigating to parent, siblings etc.) you need to create NodePaths for every node. And this is prone to errors because you have many similar and long selectors like `path.node.expression.expressions[0].type` which you can mess up easily.
Either you create NodePaths for all new nodes or you have to find out, which you need, which is mainly dependent on your operations. So every change in algorithms needs updates to the NodePath handling.
Reducing the number of operations in a traversal often allows to create new nodes only and let the next traverse create the NodePaths for them.

But, there is a second risk. If you change node relationships (mainly moving nodes to other parents), you need to correct these in the corresponding NodePaths. This is even more critical because they are cached.

I think there should be something to get NodePath handling out of the way, perhaps similar to `getParent()`, `getChild(n)` etc. (btw. I would keep `getPatent()` because of consistency).
If I didn't miss something, the NodePath module is not thought to be public. I have to require "regexp-tree/src/traverse/node-path", right?

May be creating NodePaths should become more automatic.
I am aware of the problem that parent relation can only be created bottom up.

Perhaps a createPaths for a sub-tree could solve this. Possible options could be: "forcePaths" and/or "checkPaths". When you create a new node sub-tree (often also including already existing nodes) you could call this function to (force) create the NodePaths for all the nodes and optionally check existing paths in case they got a new parent.

It turns out, that traversing is more frequent than I thought initially. I now have a lot of traversals for a relatively simple transformation.

Finally: why do we need NodePaths?
Having two dependent parallel structures often makes more trouble.
Without having read about this particular case, I see following reasons for not using parent members directly in nodes and a set of methods to handle node relationships:
* clean object oriented design (what ever that means exactly)
* clean appearance of the tree in JSON etc.
* allowing a node to be in multiple parents
* garbage collection issues with dependency cycles
I think some of these can be solved in other ways (real objects, weak references) or their value could be reconsidered (cleanness) or they are invalid for this usage (multiple parents).
My last adventure made me feel, I would prefer less complexity and cleaner handling in real use-cases over more theoretical cleanness. A simpler design (KISS) could eventually make usage much more intuitive.
Please note, I'm usually a defender of clean design principles, but sometimes the costs are much too high.

**Comment by DmitrySoshnikov** [comment]
@hg42, all good point, thanks. I'm gonna come back to revisit them a bit later (got a bit busy on main work, this is mostly my side project).

> Finally: why do we need NodePaths?

When [we were thinking](https://github.com/DmitrySoshnikov/regexp-tree/issues/7#issuecomment-291004872) about the API for transform/traverse, there was an option to traverse just nodes, and pass an API object as a parameter. In this case, a handler would be a function, which accepts API as a parameter, and returns an actual handler object:

There might pros, and cons in that approach too. E.g. for sibling nodes, you'll have to pass index explicitly to API methods:

```js
api.getNextSibling(node, parent, index);
```

The actual reasons for `NodePath` were:

- Providing API methods, which I didn't want to exposed on actual nodes (following Mozilla Parser API, where AST nodes are just JSON-compatible plain objects)

- Abstraction of the index/prop, etc.

- Convenient access to the whole chain of parents (not just one which is passed as a parameter).

- Familiarity for developers in another tool (e.g. Babel - transformer for JavaScript code), which also uses `NodePath` abstraction with similar API (like `path.parentPath`, etc).

Perhaps, we can revisit the approach with passing API as a parameter, as long as we think that boilerplate with passing explicit indexes, will be acceptable. To repeat the example from the link:

```js
const regexpTree = require('regexp-tree');

// The `api` parameter received can have methods:
//
// api.removeFromParent(parent, index)
// api.getParent(node)
// api.getNextSibling(node, parent, index)
// api.traverse(newNode)
// etc
//
// As you can see more parameters should be explicitly passed
// and NodPath incapsulates them.

const re = regexpTree.transform('/[a-z]{1,}/',
  function(api) { return {
    // Handle "Quantifier" node type,
    // transforming `{1,}` quantifier to `+`.
    onQuantifier(node) {
      // {1,} -> +
      if (
        node.type === 'Range' &&
        node.from === 1 &&
        !node.to
      ) {
        api.replaceNode(node, {
          type: 'Quantifier',
          kind: '+',
          loc: {
            start: node.loc.start,
            end: node.loc.start + 1,
          },
        });
      }
    },
  },
});

console.log(re.toRegExp()); // /[a-z]+/
```

At this point I think we'll just need to have 2-pass traversal if going to keep `NodePath`: at first pass all the `NodePath`s wrappers are created, and on the second the actual handling begins.

**Comment by hg42** [comment]
I read a bit about javascript's gc, cyclic dependency should be no problem (I was influenced by python using reference counting, javascript uses mark and sweep).

The api parameter shouldn't help, because it doesn't hold index, prop and parent.
It could hold these for the single node like an iterator object, but this is not sufficient for tree reordering (only one level).

I was thinking of storing the parent in the node itself (which you don't want, JSON etc.).
But this does not solve fast access to the reference in the children array. Without index you have to search for the node in the children array. As a consequence the index (and prop) has to be stored, too. So NodePath data would be part of the node.

But even this would not solve the main problem:
the parent reference is not managed when reordering the tree.

So, IMHO we should manage the parent relation and NodePath creation automatically in tree reordering operations.

Reordering is navigating and assigning nodes.
All navigation methods already manage the NodePath.

In another post I misinterpreted the "replace" as replacing the child of the node.
But it replaces the node of the path.
Then the path's parent does not change and the path also fits for the new Node.
*BUT*, the path is registered for the old node! it should be moved in the registry or added again.

You have already created another operation `insertChildAt`, which manages NodePaths by calling `_rebuildIndex`.

(btw. I think, you could add an `index=0` parameter to `_rebuildIndex` to start rebuilding from the insert position)

Still missing:
* `setChildAt(node, index, prop)` same as insertChildAt without inserting
* `setChild(node)` for single child
* `appendChild(node)` can call `setChild(length, node)`

perhaps the setChild* should be combined into one function:
`setChild(node, index=0, prop=null)`

`..., prop, index` in getForNode parameter list is not consistent with insertChildAt.
If prop is optional (because it could be derived from the object like in getChild) it should be at the end anyway.

(I already have implemented some of these, pull request is on the way)

**Comment by DmitrySoshnikov** [comment]
> BUT, the path is registered for the old node! it should be moved in the registry or added again.

Right, and good catch! We should fix `replace` and take care of it. Although.. perhaps this currently "magically works" (haven't checked), and a new `NodePath` is created when this new node is trying to be accessed from some method.

> Still missing:

Yeah, we can add those methods. I can take a look later, or feel free to send a PR as well :)

**Comment by hg42** [comment]
if the node is totally new, the corresponding NodePath would be created, **BUT** if the node already has a NodePath it will be reused and has the wrong parent!
I already changed `getForNode` to force given parameters on the cached path.
I also had code to create/change NodePaths in replace, but it didn't work because getForNode only used the cached path as it is. I think I'll add that to setChild/appendChild, ok?
